### ISSUE TITLE
Async support?

### ISSUE THREAD

**Issue by Kixunil** [bitcoindevkit/rust-electrum-client#96]
As a networking-related dependency, this would be a great candidate for making it async.

**Comment by oleonardolima** [comment]
As I mentioned at https://github.com/bitcoindevkit/bdk/issues/66 I guess it would also make it possible to add Tor support using arti-client.

**Comment by Kixunil** [comment]
@oleonardolima I think just make it possible to switch up any transport layer using a trait. Probably `tokio::AsyncRead` etc.

**Comment by oleonardolima** [comment]
Good, that makes sense. I'll try to take a look at it after adding the support on `rust-esplora-client`

**Comment by Kixunil** [comment]
@oleonardolima FYI I hand-wrote an async client with a few methods that were required for my project. I may be able to publish it as it may serve as a starting point.

**Comment by oleonardolima** [comment]
@Kixunil Cool! Thanks, that would be really helpful and a great starting point.

**Comment by RCasatta** [comment]
I just finished here thinking to trying to use [arti_client](https://docs.rs/arti-client/latest/arti_client/#connecting-to-tor) with rust-electrum-client but noting that is async based while this is not.

Even some temporary hack like [SyncIoBridge](https://docs.rs/tokio-util/latest/tokio_util/io/struct.SyncIoBridge.html) seems not easily applicable here, because the stream is not handled splitted (reparated Read/Write)

**Comment by Kixunil** [comment]
@RCasatta yeah, IIRC that's why I ended up writing my own and as I did I found that the whole thing is quite involved.

@oleonardolima sorry to keep you waiting this long I will try soon.

**Comment by oleonardolima** [comment]
@Kixunil No worries, I'm still working on `fedimint` and `rust-esplora-client` tor support at the moment, but let me know when you do it.

**Comment by Kixunil** [comment]
It's out now: https://github.com/Kixunil/async-electrum-client

**Comment by oleonardolima** [comment]
update: I'll probably start working on this during this week, using the @Kixunil work as a starting point.

Do you have any other approach in mind @RCasatta ?

**Comment by RCasatta** [comment]
> Do you have any other approach in mind @RCasatta ?

No

**Comment by jurvis** [comment]
@oleonardolima  curious, are you still actively working on this?

**Comment by oleonardolima** [comment]
@jurvis I'm not currently working on this, other issues took priority, although it's still on my list. Is it a pressing issue/feature that you need?

**Comment by jurvis** [comment]
@oleonardolima not pressing, just asking! do you have a draft branch going somewhere?
